### PR TITLE
Add bulk_get_node function

### LINKED ISSUE #14110: lua api: Function to get adjacent nodes.
### Problem

I think it would be useful to have a Lua API to check for adjacent nodes of a given location. This is different to `find_node_near` as the blocks found would be directly touching the node where the `find_node_near` function will get nodes that are diagonal to it.

### Solutions

Adds a new Lua API function to find touching blocks.

### Alternatives

Add an option to `find_node_near` or `find_nodes_in_area` to only search for touching blocks.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sfence**
- Add a Lua function for getting multiple nodes at one call.
- Fixes  #14110.

## To do

This PR is a Ready for Review.

## How to test

Run devgame `/unittests` command.

### CODE REVIEW

File: doc/lua_api.md [hunk 1]
```diff
@@ -5869,6 +5869,14 @@ Environment access
       returns `{name="ignore", param1=0, param2=0}` for unloaded areas.
 * `minetest.get_node_or_nil(pos)`
     * Same as `get_node` but returns `nil` for unloaded areas.
+* `minetest.bulk_get_node({pos1, pos2, pos3, ...})`
+    * Get nodes from all positions set in the first argument.
+    * e.g. `minetest.bulk_get_node({{x=0, y=1, z=1}, {x=1, y=2, z=2}})`
+    * For node specification or position syntax see `minetest.get_node` call
+    * Faster than `set_node` due to single call, but still considerably slower
```
Comment (AFCMS) [inline doc/lua_api.md:5876]:
```suggestion
    * Faster than `get_node` due to single call, but still considerably slower
```

**Comment by rubenwardy** [comment]
Please may you include benchmarks, as that is the justification for this feature

File: doc/lua_api.md [hunk 2]
```diff
@@ -5869,6 +5869,14 @@ Environment access
       returns `{name="ignore", param1=0, param2=0}` for unloaded areas.
 * `minetest.get_node_or_nil(pos)`
     * Same as `get_node` but returns `nil` for unloaded areas.
+* `minetest.bulk_get_node({pos1, pos2, pos3, ...})`
+    * Get nodes from all positions set in the first argument.
+    * e.g. `minetest.bulk_get_node({{x=0, y=1, z=1}, {x=1, y=2, z=2}})`
+    * For node specification or position syntax see `minetest.get_node` call
```
Comment (sfan5) [inline doc/lua_api.md:5875]: i think it's obvious
```suggestion
```

File: doc/lua_api.md [hunk 3]
```diff
@@ -5869,6 +5869,14 @@ Environment access
...
+    * Faster than `set_node` due to single call, but still considerably slower
+      than Lua Voxel Manipulators (LVM) for large numbers of nodes.
+      Unlike LVMs, this will call node callbacks. It also allows getting nodes
+      from spread out positions which would cause LVMs to waste memory.
```
Comment (sfan5) [inline doc/lua_api.md:5879]:
```suggestion
      It allows getting nodes from spread out positions
      which would cause LVMs to waste memory.
```

File: games/devtest/mods/unittests/bulk.lua [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+		table.insert(positions, vector.add(pos, vector.new(0, n, 0)))
+	end
+	local nodes = minetest.bulk_get_node(positions)
+
+	minetest.bulk_set_node(positions, {name="basenodes:stone"})
+
+	for index, set_pos in pairs(positions) do
+		minetest.set_node(set_pos, nodes[index])
+	end
+end
```
Comment (sfan5) [inline games/devtest/mods/unittests/bulk.lua:15]: this test is pointless, it doesn't have any asserts

File: doc/lua_api.md (hunk 3 above)
Comment (sfan5) [inline doc/lua_api.md:5879]: also missing explanation of the return value

File: src/script/lua_api/l_env.cpp [hunk 5]
```diff
@@ -359,6 +359,39 @@ int ModApiEnv::l_get_node_or_nil(lua_State *L)
 	return 1;
 }
 
+// bulk_get_node([pos1, pos2, ...])
+// pos = {x=num, y=num, z=num}
```
Comment (sfan5) [inline src/script/lua_api/l_env.cpp:363]: IMO these comments are largely pointless because we have separate docs and they get out-of-date fast
```suggestion
// bulk_get_node(pos_list)
```

File: src/script/lua_api/l_env.cpp [hunk 6]
```diff
@@ -359,6 +359,39 @@ int ModApiEnv::l_get_node_or_nil(lua_State *L)
...
+int ModApiEnv::l_bulk_get_node(lua_State *L)
+{
+	GET_ENV_PTR;
+
+	// parameters
+	if (!lua_istable(L, 1)) {
+		return 0;
```
Comment (sfan5) [inline src/script/lua_api/l_env.cpp:370]: should probably be an error?

File: src/script/lua_api/l_env.cpp [hunk 7]
```diff
@@ -359,6 +359,39 @@ int ModApiEnv::l_get_node_or_nil(lua_State *L)
...
+	}
+
+	s32 len = lua_objlen(L, 1);
+
+	// Do it
+	std::vector<MapNode> nodes;
+
+	for (s32 i = 1; i <= len; i++) {
+		lua_rawgeti(L, 1, i);
+		nodes.push_back(env->getMap().getNode(read_v3s16(L, -1)));
```
Comment (sfan5) [inline src/script/lua_api/l_env.cpp:380]: it's a bit more complicated but you don't need to store a copy of all the nodes read

File: src/script/lua_api/l_env.cpp [hunk 8]
```diff
@@ -359,6 +359,39 @@ int ModApiEnv::l_get_node_or_nil(lua_State *L)
...
+		lua_pop(L, 1);
+	}
+
+	// create table for return data
+	lua_newtable(L);
```
Comment (sfan5) [inline src/script/lua_api/l_env.cpp:385]:
```suggestion
	lua_createtable(L, len, 0);
```

File: src/script/lua_api/l_env.cpp (hunk 7 above)
Comment (sfence) [reply src/script/lua_api/l_env.cpp:380]: Yes. But in that case, I will need to push and pop data from the Lua stack simultaneously.
Is there some example of how to do it?

File: src/script/lua_api/l_env.cpp (hunk 6 above)
Comment (sfence) [reply src/script/lua_api/l_env.cpp:370]: There is no error in this case in `l_bulk_set_node`.

So let's add an error message for `l_bulk_set_node` and for `l_bulk_get_node` or keep it without it both.
What is a preferred solution?

File: src/script/lua_api/l_env.cpp (hunk 5 above)
Comment (sfence) [reply src/script/lua_api/l_env.cpp:363]: This style of comment is used in many places in the `l_env.cpp` file.
So it takes send to keep it as is it or to remove/change them at all places in the file.
What solution do you prefer?

File: doc/lua_api.md (hunk 3 above)
Comment (sfence) [reply doc/lua_api.md]: Should I change this in the `bulk_set_nodes` command too?

File: doc/lua_api.md (hunk 2 above)
Comment (sfence) [reply doc/lua_api.md:5876]: Should I remove it from the `bulk_set_nodes` command too? Or keep it in both of them?

File: doc/lua_api.md (hunk 1 above)
Comment (sfence) [reply doc/lua_api.md]: fixed

File: games/devtest/mods/unittests/bulk.lua (hunk 4 above)
Comment (sfence) [reply games/devtest/mods/unittests/bulk.lua:29]: fixed

**Comment by sfence** [comment]
@rubenwardy 
Benchmark added. 665 ms vs 531 ms on my machine for actual implementation.

File: doc/lua_api.md (hunk 3 above)
Comment (sfence) [reply doc/lua_api.md]: fixed

Comment (sfan5) [reply doc/lua_api.md]: No, it's correct there. But there are no callbacks for *getting* nodes.

File: doc/lua_api.md (hunk 2 above)
Comment (sfan5) [reply doc/lua_api.md:5876]: If you want to.

File: src/script/lua_api/l_env.cpp (hunk 5 above)
Comment (sfan5) [reply src/script/lua_api/l_env.cpp:363]: Changing all of them is better left to a different PR.
I'd simply not add more to the pile and keep the comment shorter (as suggested).

File: src/script/lua_api/l_env.cpp (hunk 6 above)
Comment (sfan5) [reply src/script/lua_api/l_env.cpp:370]: Guess not then.

File: src/script/lua_api/l_env.cpp (hunk 7 above)
Comment (sfan5) [reply src/script/lua_api/l_env.cpp:380]:
```cpp
const int list = lua_gettop(L);
s32 len = lua_objlen(L, 1);

lua_createtable(L, len, 0);
const int result = lua_gettop(L);

auto &map = env->getMap();
for (s32 i = 1; i <= len; i++) {
	lua_rawgeti(L, list, i);
	auto pos = read_v3s16(L, -1);
	lua_pop(L, 1);
	pushnode(map.getNode(pos));
	lua_rawseti(L, result, i);
}

return 1;
```

should work

File: src/script/lua_api/l_env.cpp (hunk 5 above)
Comment (sfence) [reply src/script/lua_api/l_env.cpp]: ok, done.

File: src/script/lua_api/l_env.cpp (hunk 7 above)
Comment (sfence) [reply src/script/lua_api/l_env.cpp]: Thanks for your help. Done.

File: doc/lua_api.md [hunk 9]
```diff
@@ -5869,6 +5869,16 @@ Environment access
       returns `{name="ignore", param1=0, param2=0}` for unloaded areas.
 * `minetest.get_node_or_nil(pos)`
     * Same as `get_node` but returns `nil` for unloaded areas.
+* `minetest.bulk_get_node({pos1, pos2, pos3, ...})`
+    * Get nodes from all positions set in the first argument.
+    * Returns list of nodes.
```
Comment (sfan5) [inline doc/lua_api.md:5874]:
```suggestion
    * Returns list of nodes in the same order as the input list.
```

File: doc/lua_api.md [hunk 10]
```diff
@@ -5869,6 +5869,16 @@ Environment access
...
+    * e.g. `minetest.bulk_get_node({{x=0, y=1, z=1}, {x=1, y=2, z=2}})`
+    * For node specification or position syntax see `minetest.get_node` call
+    * Faster than `get_node` due to single call, but still considerably slower
+      than Lua Voxel Manipulators (LVM) for large numbers of nodes.
+      It also allows getting nodes from spread out positions which would cause
+      LVMs to waste memory.
+      For getting a cube, this is 1.3x faster than `set_node`.
```
Comment (sfan5) [inline doc/lua_api.md:5881]: it doesn't matter too much but: a cube is what? 3x3x3?

File: games/devtest/mods/benchmarks/init.lua [hunk 11]
```diff
@@ -114,4 +114,39 @@ minetest.register_chatcommand("bench_bulk_set_node", {
...
+		minetest.chat_send_player(name, "Benchmarking minetest.bulk_get_node...");
+
+		local start_time = minetest.get_us_time()
+		for i=1,#pos_list do
+			minetest.get_node(pos_list[i])
+		end
+		local middle_time = minetest.get_us_time()
+		minetest.bulk_get_node(pos_list, {name = "mapgen_stone"})
+		local end_time = minetest.get_us_time()
+		local msg = string.format("Benchmark results: minetest.set_node loop: %.2f ms; minetest.bulk_set_node: %.2f ms",
```
Comment (sfan5) [inline games/devtest/mods/benchmarks/init.lua:146]: string wrong

File: games/devtest/mods/benchmarks/init.lua [hunk 12]
```diff
@@ -114,4 +114,39 @@ minetest.register_chatcommand("bench_bulk_set_node", {
...
+		if not player then
+			return false, "No player."
+		end
+		local pos_list = {}
+		local ppos = player:get_pos()
+		local i = 1
+		for x=2,100 do
+			for y=2,100 do
+				for z=2,100 do
+					pos_list[i] = {x=ppos.x + x,y = ppos.y + y,z = ppos.z + z}
```
Comment (sfan5) [inline games/devtest/mods/benchmarks/init.lua:131]: copy-paste, please move to function to deduplicate

File: games/devtest/mods/benchmarks/init.lua [hunk 13]
```diff
@@ -114,4 +114,39 @@ minetest.register_chatcommand("bench_bulk_set_node", {
...
+					i = i + 1
+				end
+			end
+		end
 
+		minetest.chat_send_player(name, "Benchmarking minetest.bulk_get_node...");
```
Comment (sfan5) [inline games/devtest/mods/benchmarks/init.lua:137]: `;` superfluous

File: games/devtest/mods/unittests/bulk.lua [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	minetest.bulk_set_node(positions, {name="basenodes:stone"})
+
+	for _, npos in pairs(positions) do
+		local node = minetest.get_node(npos)
+		assert(node.name == "basenodes:stone")
+		assert(node.param1 == 0)
+		assert(node.param2 == 0)
+	end
+
+	for ipos, npos in pairs(positions) do
```
Comment (sfan5) [inline games/devtest/mods/unittests/bulk.lua:26]:
```suggestion
	-- put back old nodes
	for ipos, npos in pairs(positions) do
```

File: doc/lua_api.md (hunk 10 above)
Comment (sfence) [reply doc/lua_api.md]: Updated aslo in bulk_set_node.

File: games/devtest/mods/benchmarks/init.lua (hunk 12 above)
Comment (sfence) [reply games/devtest/mods/benchmarks/init.lua]: done

File: games/devtest/mods/benchmarks/init.lua (hunk 13 above)
Comment (sfence) [reply games/devtest/mods/benchmarks/init.lua]: Fixed also in bench_bulk_set_ndoe.

File: doc/lua_api.md [hunk 15]
```diff
@@ -5869,6 +5869,16 @@ Environment access
...
+* `minetest.bulk_get_node({pos1, pos2, pos3, ...})`
+    * Get nodes from all positions set in the first argument.
+    * Returns list of nodes in the same order as the input list.
+    * e.g. `minetest.bulk_get_node({{x=0, y=1, z=1}, {x=1, y=2, z=2}})`
+    * For node specification or position syntax see `minetest.get_node` call
+    * Faster than `get_node` due to single call, but still considerably slower
+      than Lua Voxel Manipulators (LVM) for large numbers of nodes.
+      It also allows getting nodes from spread out positions which would cause
+      LVMs to waste memory.
+      For getting a cube of size 99x99x99, this is 1.3x faster than `set_node`.
```
Comment (sfan5) [inline doc/lua_api.md:5881]: get_node*

File: games/devtest/mods/unittests/bulk.lua [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	minetest.bulk_set_node(positions, {name="basenodes:stone"})
+
+	for _, npos in pairs(positions) do
+		local node = minetest.get_node(npos)
+		assert(node.name == "basenodes:stone")
+		assert(node.param1 == 0)
+		assert(node.param2 == 0)
+	end
+
+	-- push back old nodes
```
Comment (sfan5) [inline games/devtest/mods/unittests/bulk.lua:26]: put*

**Review by sfan5 — APPROVED** [review]
code looks good and CI is green

**Review by appgurueu — COMMENTED** [review]
The implementation looks reasonable. Just a thought: I wonder whether we could achieve more than 1.3x (especially if we take building the input tables into account) if we change the interface a bit:

* Have you considered passing the input positions as three separate lists of `x`, `y` and `z` tables?
* Alternatively, packed even tighter, we could pass input positions as a list of hashed node positions.
* Output could be three lists of nodenames / param1s / param2s, or maybe even content IDs / param1s / param2s like voxelmanip. This could also be bitpacked in one table, though that might make extraction of the values awkward or inefficient.
* We could also consider a faster `get_node` interface which just pushes the 3 values to the stack rather than building a hash table each time.

**Comment by sfence** [comment]
@appgurueu 
Sounds like an idea for new Lua API functions.

File: games/devtest/mods/benchmarks/init.lua [hunk 17]
```diff
@@ -114,4 +120,29 @@ minetest.register_chatcommand("bench_bulk_set_node", {
...
+		minetest.chat_send_player(name, "Benchmarking minetest.bulk_get_node...")
 
+		local start_time = minetest.get_us_time()
+		for i=1,#pos_list do
+			minetest.get_node(pos_list[i])
+		end
+		local middle_time = minetest.get_us_time()
+		minetest.bulk_get_node(pos_list, {name = "mapgen_stone"})
+		local end_time = minetest.get_us_time()
+		local msg = string.format("Benchmark results: minetest.get_node loop: %.2f ms; minetest.bulk_get_node: %.2f ms",
```
Comment (SmallJoker) [inline games/devtest/mods/benchmarks/init.lua:142]: Like set_bulk_node, this is slightly biased because `get_positions_cube` is not taken into account.

**Comment by SmallJoker** [comment]
(sorry for the fuss. deleted my erroneous comment)

**Comment by appgurueu** [comment]
Okay, I finally got to benchmarking this. I'm afraid the benchmark seems to be flawed. `minetest.bulk_get_node` seems to be getting an advantage due to running second. You can verify this by swapping the two sections in the benchmark: If `minetest.get_node` goes first, it wins, and by a similar factor; I conjecture that this probably is due to caching effects (possibly both at software and hardware levels):

Screenshot from 2024-01-18 21-12-15

(Furthermore, the current code does not store `minetest.get_node` in a local variable, but this does not seem to have a major impact. The `bulk_set_node` benchmark suffers from the same flaws, though the order of operations should be less relevant there since there is a proper "warmup" phase.)

The changes I made for benchmarking:

```diff
diff --git a/games/devtest/mods/benchmarks/init.lua b/games/devtest/mods/benchmarks/init.lua
index 75d959c3d..cf4e4176e 100644
--- a/games/devtest/mods/benchmarks/init.lua
+++ b/games/devtest/mods/benchmarks/init.lua
@@ -133,15 +133,16 @@ minetest.register_chatcommand("bench_bulk_get_node", {
                minetest.chat_send_player(name, "Benchmarking minetest.bulk_get_node...")
 
                local start_time = minetest.get_us_time()
+               minetest.bulk_get_node(pos_list, {name = "mapgen_stone"})
+               local middle_time = minetest.get_us_time()
+               local get_node = minetest.get_node
                for i=1,#pos_list do
-                       minetest.get_node(pos_list[i])
+                       get_node(pos_list[i])
                end
-               local middle_time = minetest.get_us_time()
-               minetest.bulk_get_node(pos_list, {name = "mapgen_stone"})
                local end_time = minetest.get_us_time()
                local msg = string.format("Benchmark results: minetest.get_node loop: %.2f ms; minetest.bulk_get_node: %.2f ms",
-                       ((middle_time - start_time)) / 1000,
-                       ((end_time - middle_time)) / 1000
+                       ((end_time - middle_time)) / 1000,
+                       ((middle_time - start_time)) / 1000
                )
                return true, msg
        end,
```

I'm not sure that "bulking" by means of a list of node positions as parameters / nodes as return value is the right way to optimize this. This reduces the Lua - C context switch overhead, but that overhead isn't all that large, and we pay for it by having to store the nodes we return in a table. (In the future, a "bulk" API could do further "batching" optimizations, but currently this isn't being done. Hence there is a case to be made for introducing a "bulk" API both for modder convenience, and for prospective future performance improvements which that would enable.)

**TL;DR:**

* The current benchmarks are unfortunately flawed. As it stands, I see no reason to assume a significant performance improvement.
* This PR might still have merit for modder convenience and future performance improvements.

File: games/devtest/mods/benchmarks/init.lua [hunk 18]
```diff
@@ -114,4 +120,29 @@ minetest.register_chatcommand("bench_bulk_set_node", {
...
+		local pos_list = get_positions_cube(player:get_pos())
+
+		minetest.chat_send_player(name, "Benchmarking minetest.bulk_get_node...")
 
+		local start_time = minetest.get_us_time()
+		for i=1,#pos_list do
+			minetest.get_node(pos_list[i])
+		end
+		local middle_time = minetest.get_us_time()
+		minetest.bulk_get_node(pos_list, {name = "mapgen_stone"})
```
Comment (Desour) [inline games/devtest/mods/benchmarks/init.lua:140]:
```suggestion
		minetest.bulk_get_node(pos_list)
```

**Comment by sfence** [comment]
Benchmark updated.
My samples for it:
|         | get_node         | bulk_get_node    |   ration  |
| ------- | ---------------- | ---------------- | ---------------- |
| mean    | 586.799353901595 | 450.751030768249 | 1.30182587248102 |
| avarage | 590.270625       | 459.245625       | 1.28530484095521 |
|         | 584.5            | 506.81           |                  |
|         | 680.22           | 490.4            |                  |
|         | 642.22           | 374.14           |                  |
|         | 509.17           | 542.02           |                  |
|         | 551.68           | 508.02           |                  |
|         | 557.92           | 579.47           |                  |
|         | 507.39           | 532.36           |                  |
|         | 649.84           | 379.67           |                  |
|         | 615.15           | 370.21           |                  |
|         | 492.51           | 491.43           |                  |
|         | 620.45           | 369.11           |                  |
|         | 499.58           | 674.21           |                  |
|         | 707.78           | 378.6            |                  |
|         | 603.2            | 376.1            |                  |
|         | 597.74           | 377.59           |                  |
|         | 624.98           | 397.79           |                  |

**Comment by sfence** [comment]
Storing get_node in a local variable does not look to make a significant effect on my machine:
|         | get_node         | bulk_get_node    | ration           |
| ------- | ---------------- | ---------------- | ---------------- |
| mean    | 547.849938348686 | 425.677725965754 | 1.28700635464483 |
| avarage | 553.707777777778 | 433.313333333333 | 1.27784615700212 |
|         | 638.78           | 428.49           |                  |
|         | 709.79           | 416.42           |                  |
|         | 470.31           | 644.37           |                  |
|         | 592.57           | 374.84           |                  |
|         | 586.63           | 346.03           |                  |
|         | 469              | 478.32           |                  |
|         | 479.28           | 345.91           |                  |
|         | 473.51           | 397.21           |                  |
|         | 563.5            | 468.23           |                  |

**Comment by Desour** [comment]
FYI, most of the overhead of `get_node` and `get_node_bulk` comes from reading vectors and pushing nodes, because that does C++->Lua calls. If you do something like this for example, you get something faster than `get_node_bulk`: https://github.com/Desour/minetest/commit/c40ccd25ae21e7517ea068cceb68eaaf89fcdf19
#12438 should be adopted.

**Comment by sfan5** [comment]
So are we doing this or not?

**Comment by appgurueu** [comment]
I think we can add this as a convenience feature (which might lend itself better to future optimization work than `get_node`), but we should probably remove the performance hint / advice from the docs for now.

**Comment by grorp** [comment]
If we add this, I think we should go a step further: We should explicitly state that there is no performance benefit to avoid modders doing premature optimisation.

**Comment by sfence** [comment]
Removed performance hint from doc.

**Comment by Desour** [comment]
Rebase needed.

Also, with #14384 merged, this feature is now somewhat obsolete. @sfence Thoughts?

**Comment by rubenwardy** [comment]
I don't think this offers a meaningful usability improvement, so without the performance improvement I'd say this is not needed

**Comment by Zughy** [comment]
@sfence do you still see a reason for this PR or a way to make it faster? See today's IRC log https://irc.minetest.net/minetest-dev/2024-03-03#i_6156853

**Comment by sfence** [comment]
I have created this PR to fix #14110.  If it is no longer interesting, please close #14110 as won't add.

This PR has performance benefits in comparison with `core.get_node` in most cases, as the benchmark shows. The problem is, that the performance benefits look to be different on different machines.

**Comment by sfence** [comment]
From the API side view, it can be nice to have a complement function to the `core.bulk_set_node` function.

But at all performance benefits of #14384 are much better.

**Comment by sfan5** [comment]
Considering this as rejected, then.
